### PR TITLE
document for logger folder

### PR DESCRIPTION
**Pull request by MrLYG**
document for files in logger folder

### CODE REVIEW

File: ablator/modules/loggers/file.py [hunk 1]
```diff
@@ -15,45 +28,116 @@ def __init__(
...
         if self.path is not None:
             with open(self.path, "a", encoding="utf-8") as f:
                 f.write(f"{msg}\n")
 
     def _print(self, msg: str, verbose=False):
+        """ Print a message to the console.
+
+        Parameters
+        ----------
+        msg : str
```
Comment (hjzccc) [inline ablator/modules/loggers/file.py:68]: lack parameter verbose

File: ablator/modules/loggers/file.py [hunk 2]
```diff
@@ -15,45 +28,116 @@ def __init__(
...
         if self.verbose or verbose:
             print(msg)
 
     def info(self, msg: str, verbose=False):
+        """Log an info message.
+
+        Parameters
+        ----------
+        msg : str
+            The message to log.
```
Comment (hjzccc) [inline ablator/modules/loggers/file.py:83]: ditto

File: ablator/modules/loggers/file.py [hunk 3]
```diff
@@ -15,45 +28,116 @@ def __init__(
...
+        """
         self(msg, verbose)
 
     def warn(self, msg: str, verbose=True):
+        """Log a warning message.
+
+        Parameters
+        ----------
```
Comment (hjzccc) [inline ablator/modules/loggers/file.py:93]: ditto

File: ablator/modules/loggers/file.py [hunk 4]
```diff
@@ -15,45 +28,116 @@ def __init__(
...
         msg = f"{FileLogger.FAIL}{msg}{FileLogger.ENDC}"
         self(msg, True)
 
     def __call__(self, msg: str, verbose=True):
+        """Log a message.
+
+        Parameters
+        ----------
+        msg : str
+            The message to log.
```
Comment (hjzccc) [inline ablator/modules/loggers/file.py:119]: ditto

File: ablator/modules/loggers/main.py [hunk 5]
```diff
@@ -21,6 +21,28 @@ class DuplicateRunError(Exception):
 
 
 class SummaryLogger:
+    """
+    A logger for training and evaluation summary.
+
+    Attributes
+    ----------
```
Comment (hjzccc) [inline ablator/modules/loggers/main.py:28]: lack some attributes here, ie. 
```Python
        self.log_iteration: int = 0
        self.checkpoint_iteration: dict[str, dict[str, int]] = {}
        self.log_file_path: Path | None = None
        self.dashboard: LoggerBase | None = None
        self.model_dir: Path | None = None
        self.result_json_path: Path | None = None
```

File: ablator/modules/loggers/main.py [hunk 6]
```diff
@@ -106,11 +148,34 @@ def _update_metadata(self):
...
+        LoggerBase | None
+            A TensorboardLogger.
+        """
         if run_config is None or not run_config.tensorboard:
             return None
         return TensorboardLogger(summary_dir.joinpath("tensorboard"))
 
     def _write_config(self, run_config: RunConfig):
+        """
+        Write the run configuration to the model directory.
```
Comment (hjzccc) [inline ablator/modules/loggers/main.py]: write to model diretory and to dashboard

File: ablator/modules/loggers/main.py [hunk 7]
```diff
@@ -189,6 +282,25 @@ def checkpoint(
         itr: int | None = None,
         is_best: bool = False,
     ):
+        """Save a checkpoint.
```
Comment (hjzccc) [inline ablator/modules/loggers/main.py]: save a checkpoint and update the checkpoint iteration

File: ablator/modules/loggers/main.py [hunk 8]
```diff
@@ -164,6 +248,15 @@ def update(
         metrics: Union[TrainMetrics, dict],
         itr: Optional[int] = None,
     ):
+        """ Update the dashboard with the given metrics.
```
Comment (hjzccc) [inline ablator/modules/loggers/main.py:266]: it also writes some metrics to json files and update the current metadata(log_iteration)

**Review by hjzccc — COMMENTED** [review]
looks good

**Comment by MrLYG** [comment]
> looks good

Good work, Thank you!

**Review by hjzccc — APPROVED** [review]

File: ablator/modules/loggers/main.py [hunk 9]
```diff
@@ -164,6 +262,16 @@ def update(
...
+        write some metrics to json files and update the current metadata(log_iteration)
```
Comment (hieuchi911) [inline ablator/modules/loggers/main.py:267]: - If `itr` is passed to this method, `AssertionError` will be raised if it's less than `self.log_iteration`, aka rewriting history, which is not allowed
- Maybe mention that `self.log_iteration` is increased by 1 every time `update()` is called while training models

File: ablator/modules/loggers/main.py [hunk 10]
```diff
@@ -21,6 +21,42 @@ class DuplicateRunError(Exception):
...
+        List of checkpoint directory names.
+    CHKPT_DIR_VALUES : list[str]
+        List of checkpoint directory values.
+    CHKPT_DIRS : dict[str, Path]
+        Dictionary containing checkpoint directories.
+    keep_n_checkpoints : int
+        Number of checkpoints to keep.
+    log_iteration : int
+        Current log iteration.
+    checkpoint_iteration : dict[str, dict[str, int]]
```
Comment (hieuchi911) [inline ablator/modules/loggers/main.py:49]: This is quite general, maybe clarify how it's used to save the checkpoints (directory name, file name etc. in `checkpoint()`)

File: ablator/modules/loggers/main.py [hunk 11]
```diff
@@ -189,6 +297,25 @@ def checkpoint(
         itr: int | None = None,
         is_best: bool = False,
     ):
+        """Save a checkpoint and update the checkpoint iteration
```
Comment (hieuchi911) [inline ablator/modules/loggers/main.py:311]: - Specify how it's constructing checkpoint folder names and checkpoint names? i.e directory name (`best` if the trained model achieves the best performance, else `recent`)
- AssertionError is rasied when the passed param `itr` is less than the latest iteration associated with the checkpoint
- Should we mention how `self.checkpoint_iteration` is increased by 1 each time a checkpoint is saved, or how it's updated to `itr` if AssertionError above is not raised?

File: ablator/modules/loggers/main.py [hunk 12]
```diff
@@ -217,17 +344,55 @@ def checkpoint(
...
+        """
         if self.model_dir is None:
             return
         for chkpt_dir in self.CHKPT_DIR_VALUES:
             dir_path = self.model_dir.joinpath(chkpt_dir)
             futils.clean_checkpoints(dir_path, keep_n_checkpoints)
 
     def info(self, *args, **kwargs):
+        """
+        Log an info message using the logger.
```
Comment (hieuchi911) [inline ablator/modules/loggers/main.py]: I think we should clarify that this logs info to files and to console, not tensorboard

File: ablator/modules/loggers/main.py [hunk 13]
```diff
@@ -217,17 +344,55 @@ def checkpoint(
...
+        *args
+            Positional arguments passed to the logger's info method.
+        **kwargs
+            Keyword arguments passed to the logger's info method.
+        """
         self.logger.info(*args, **kwargs)
 
     def warn(self, *args, **kwargs):
+        """
+        Log a warning message using the logger.
```
Comment (hieuchi911) [inline ablator/modules/loggers/main.py]: I think we should clarify that this logs info to files and to console, not tensorboard

File: ablator/modules/loggers/main.py [hunk 14]
```diff
@@ -217,17 +344,55 @@ def checkpoint(
...
+        *args
+            Positional arguments passed to the logger's warn method.
+        **kwargs
+            Keyword arguments passed to the logger's warn method.
+        """
         self.logger.warn(*args, **kwargs)
 
     def error(self, *args, **kwargs):
+        """
+        Log an error message using the logger.
```
Comment (hieuchi911) [inline ablator/modules/loggers/main.py]: I think we should clarify that this logs info to files and to console, not tensorboard

File: ablator/modules/loggers/main.py (hunk 11 above)
Comment (MrLYG) [reply ablator/modules/loggers/main.py:311]: Should we mention how self.checkpoint_iteration is increased by 1 each time a checkpoint is saved, or how it's updated to itr if AssertionError above is not raised?

I think yes, giving more explanation is not a bad thing.

**Review by hieuchi911 — APPROVED** [review]
Looks good

**Review by fostiropoulos — APPROVED** [review]

**Pull request merged**